samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -272,4 +285,44 @@ StreamProcessor createStreamProcessor(
           taskFactory.getClass().getCanonicalName()));
     }
   }
+
+  /**
+   * Generates a unique latch id that is consistent across processors within the same application lifecycle.
+   * Each {@code StreamSpec} has an id that is unique and is used for hashcode computation.
+   *
+   * @param intStreams list of {@link StreamSpec}s
+   * @return latch id
+   */
+  private String generateLatchId(List<StreamSpec> intStreams) {
+    return String.valueOf(Objects.hashCode(intStreams.stream()
+        .map(StreamSpec::getId)
+        .collect(Collectors.toList())));
+  }
+
+  /**
+   * In order to fix SAMZA-1385, we are limiting the scope of coordination util within stream creation phase and destroying
+   * the coordination util right after. By closing the zk connection, we clean up the ephemeral node used for leader election.
+   * It results in a leader election whenever a new process joins and also creates intermediate streams.
```

Comment:
minor: "it creates the following issues when a new process joins after the ephemeral node is gone"